### PR TITLE
Follow up the while signature change.

### PR DESCRIPTION
**Pull request by trentlo**
- Assert that the input and output resources are in the same order to make the
   expected signature of the xla computation more straightforward.

### CODE REVIEW

**Comment by trentlo** [comment]
@sanjoy, this is the (minor) cleanup I promised you a while ago. Please help to take a look.

@jlebar @hawkinsp

**Review by sanjoy — APPROVED** [review]
lgtm

**Review by sanjoy — CHANGES_REQUESTED** [review]
Hi Trent,

Internally this fails tensorflow/compiler/tests:eager_test_cpu (not sure how it passed presubmit in github, doing an OSS build now).

The problem is that TF eager will sometimes compile regular function calls with XLA, and in such cases the resource outputs may not all be grouped together in the end.  However `XlaComputationLaunchContext::PopulateOutputs` will do the right thing for them.

Do you mind moving this `TF_RET_CHECK` to happen only for while loops since that is the only case where it is guaranteed that this invariant will hold?

**Comment by trentlo** [comment]
Understood. Let me see if I can place it in a better place.

**Comment by trentlo** [comment]
Done. Please help to take a look.

Although I think it should not affect the change, I don't fully understand the purpose of the token inputs/outputs. How are they generated, stack access ops?

File: tensorflow/compiler/tf2xla/kernels/while_op.cc [hunk 1]
```diff
@@ -37,6 +37,30 @@ namespace tensorflow {
...
+                                   const NameAttrList& body_name_attr) {
+  const FunctionBody* body;
+  TF_RETURN_IF_ERROR(ctx->compiler()->FindFunctionBody(body_name_attr, &body));
+  bool has_seen_resource = false;
+  for (int i = 0; i < body->arg_types.size(); i++) {
+    DataType arg_type = body->arg_types[i];
+    if (has_seen_resource) {
+      if (arg_type != DT_RESOURCE) {
+          return errors::InvalidArgument(
+              "Expect input resources are grouped in the end of while body ",
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/while_op.cc]: Let's also use `i` in the error message.  If you think that's not useful then please change the loop to a range for loop.

File: tensorflow/compiler/tf2xla/kernels/while_op.cc [hunk 2]
```diff
@@ -37,6 +37,30 @@ namespace tensorflow {
 
 namespace {
 
+// Verify that input resources are grouped in the end.
+Status VerifyBodyFunctionSignature(XlaOpKernelContext* ctx,
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/while_op.cc]: Why not `VerifyResourcesArgsGroupedAtEnd`

**Review by sanjoy — CHANGES_REQUESTED** [review]
Minor nits, otherwise looks great!

**Comment by sanjoy** [comment]
> Done. Please help to take a look.
> 
> Although I think it should not affect the change, I don't fully understand the purpose of the token inputs/outputs. How are they generated, stack access ops?

There are some internal users for this, but I'm not sure if there are any external uses.

Roughly: `TOKEN` values let us sequence side effecting operations like `outfeed` and `infeed` in XLA, and we might need to thread these values through function calls, functional while loops etc.

**Comment by trentlo** [comment]
Revised accordingly.

Thanks for the explanations on `token`.

**Review by sanjoy — APPROVED** [review]
lgtm

**Comment by rthadur** [comment]
This is merged by the commit: `5205ef9` , so closing this PR

**Pull request closed without merging**